rollup/eventwatcher/eventwatcher.go

```diff
@@ -0,0 +1,397 @@
+		}
+		chunkRanges = append(chunkRanges, &rawdb.ChunkRange{
+			StartBlockNumber: startBlockNumber,
+			EndBlockNumber:   startBlockNumber + numBlocks - 1,
+		})
+		startBlockNumber += numBlocks
+	}
+
+	return chunkRanges, nil
+}
+
+// getBlocksInRange retrieves blocks from the blockchain within specified chunk ranges.
+func (s *EventWatcher) getBlocksInRange(chunkRanges []*rawdb.ChunkRange) ([]*types.Block, error) {
+	var blocks []*types.Block
+
+	for _, chunkRange := range chunkRanges {
+		for i := chunkRange.StartBlockNumber; i <= chunkRange.EndBlockNumber; i++ {
+			block := s.bc.GetBlockByNumber(i)
+			if block == nil {
+				return nil, fmt.Errorf("failed to get block by number: %v", i)
+			}
+			blocks = append(blocks, block)
+		}
+	}
+
+	return blocks, nil
+}
+
+// convertBlocksToChunks processes and groups blocks into chunks based on provided chunk ranges.
+func (s *EventWatcher) convertBlocksToChunks(blocks []*types.Block, chunkRanges []*rawdb.ChunkRange) ([]*Chunk, error) {
+	if len(blocks) == 0 {
+		return nil, fmt.Errorf("invalid arg: empty blocks")
+	}
+
+	wrappedBlocks := make([]*WrappedBlock, len(blocks))
+	for i, block := range blocks {
+		txData := txsToTxsData(block.Transactions())
+		state, err := s.bc.StateAt(block.Hash())
+		if err != nil {
+			return nil, fmt.Errorf("failed to get block state, block: %v, err: %w", block.Hash().Hex(), err)
+		}
+		withdrawRoot := withdrawtrie.ReadWTRSlot(rcfg.L2MessageQueueAddress, state)
+		wrappedBlocks[i] = &WrappedBlock{
+			Header:       block.Header(),
+			Transactions: txData,
+			WithdrawRoot: withdrawRoot,
+		}
+	}
+
+	minBlockNumber := blocks[0].Header().Number.Uint64()
+	var chunks []*Chunk
+	for _, cr := range chunkRanges {
+		start, end := cr.StartBlockNumber-minBlockNumber, cr.EndBlockNumber-minBlockNumber
+		// ensure start and end are within valid range.
+		if start < 0 || end >= uint64(len(wrappedBlocks)) || start > end {
+			return nil, fmt.Errorf("invalid chunk range, start: %v, end: %v, block len: %v", start, end, len(wrappedBlocks))
+		}
+		chunk := &Chunk{
+			Blocks: wrappedBlocks[start : end+1],
+		}
+		chunks = append(chunks, chunk)
+	}
+
+	return chunks, nil
+}
+
+// reconciliation verifies the consistency between l1 contract and l2 node data.
+// crash once the consistency check fails.
+func reconciliation(batchIndex uint64, batchHash common.Hash, stateRoot common.Hash, withdrawRoot common.Hash, parentBatchMeta *rawdb.FinalizedBatchMeta, chunks []*Chunk) error {
+	if len(chunks) == 0 {
+		return fmt.Errorf("invalid arg: length of chunks is 0")
+	}
+	lastChunk := chunks[len(chunks)-1]
+	if len(lastChunk.Blocks) == 0 {
+		return fmt.Errorf("invalid arg: block number of last chunk is 0")
+	}
+	lastBlock := lastChunk.Blocks[len(lastChunk.Blocks)-1]
+	localWithdrawRoot := lastBlock.WithdrawRoot
+	if localWithdrawRoot != withdrawRoot {
+		log.Crit("Withdraw root mismatch", "l1 withdraw root", withdrawRoot.Hex(), "l2 withdraw root", localWithdrawRoot.Hex())
```

Comment:
Actually, maybe crashing like this is not great, it might leave the DB in an inconsistent state. Two options to consider:
1. Graceful shutdown, send a shutdown signal to the node and wait for services to stop
2. Print a visible error and keep running.

Reply:
have considered "inconsistent db state", it seems OK since we only care about `WriteFinalizedL2BlockNumber`.

Reply:
I don't mean this part of the DB but other processes (like state and ledger storage). Crashing on DB error seems acceptable since those are not expected (unless the DB is already corrupted). But crashing here might cause DB corruption on state etc. entries.

Reply:
got it.

Reply:
changed to `node.Close()` in [01e17d6](https://github.com/scroll-tech/go-ethereum/pull/515/commits/01e17d6c8a179ed5c935eaca1b0341e08f9de234)